File: core/fvm_operators.py
"""
FVM 이산화 연산자: 확산, 대류, 소스, 시간항.

선형 시스템 Ax = b 형태의 행렬/벡터 기여분을 반환한다.
"""

import numpy as np
from scipy import sparse
from mesh.mesh_reader import FVMesh
from core.fields import ScalarField, VectorField


class FVMSystem:
    """FVM 이산화 결과를 저장하는 선형 시스템."""

    def __init__(self, n: int):
        self.n = n
        # COO 형식 데이터
        self.rows: list = []
        self.cols: list = []
        self.vals: list = []
        self.rhs = np.zeros(n, dtype=np.float64)
        # 대각 계수 (SIMPLE 알고리즘에서 a_P 접근용)
        self.diag = np.zeros(n, dtype=np.float64)

    def add_diagonal(self, i: int, val: float):
        """대각 요소에 기여."""
        self.rows.append(i)
        self.cols.append(i)
        self.vals.append(val)
        self.diag[i] += val

    def add_off_diagonal(self, i: int, j: int, val: float):
        """비대각 요소에 기여."""
        self.rows.append(i)
        self.cols.append(j)
        self.vals.append(val)

    def add_source(self, i: int, val: float):
        """우변 벡터에 기여."""
        self.rhs[i] += val

    def to_sparse(self) -> sparse.csr_matrix:
        """CSR 희소 행렬 반환."""
        return sparse.coo_matrix(
            (self.vals, (self.rows, self.cols)),
            shape=(self.n, self.n)
        ).tocsr()

    def reset(self):
        """시스템 초기화."""
        self.rows.clear()
        self.cols.clear()
        self.vals.clear()
        self.rhs[:] = 0.0
        self.diag[:] = 0.0


def diffusion_operator(mesh: FVMesh, gamma: ScalarField, system: FVMSystem):
    """
    확산항 이산화: ∫∇·(Γ∇φ)dV ≈ Σ_f Γ_f·(φ_N - φ_P)/d_PN · A_f

    Parameters
    ----------
    mesh : FVMesh
    gamma : 확산 계수 필드 (셀 중심)
    system : 기여할 선형 시스템
    """
    for fid, face in enumerate(mesh.faces):
        owner = face.owner

        if face.neighbour >= 0:
            # 내부면
            neighbour = face.neighbour
            xO = mesh.cells[owner].center
            xN = mesh.cells[neighbour].center
            d_PN = np.linalg.norm(xN - xO)

            if d_PN < 1e-30:
                continue

            # 면에서의 확산 계수 (조화 평균)
            gamma_O = gamma.values[owner]
            gamma_N = gamma.values[neighbour]
            if gamma_O + gamma_N > 1e-30:
                gamma_f = 2.0 * gamma_O * gamma_N / (gamma_O + gamma_N)
            else:
                gamma_f = 0.0

            coeff = gamma_f * face.area / d_PN

            system.add_diagonal(owner, coeff)
            system.add_diagonal(neighbour, coeff)
            system.add_off_diagonal(owner, neighbour, -coeff)
            system.add_off_diagonal(neighbour, owner, -coeff)

        else:
            # 경계면: apply_boundary_conditions에서 BC 유형에 따라 처리
            # (Dirichlet → aP += D, b += D*phi_b; zero_gradient → 기여 없음)
            pass


def convection_operator_upwind(mesh: FVMesh, mass_flux: np.ndarray, system: FVMSystem):
    """
    대류항 이산화 (1차 Upwind): ∫∇·(ρuφ)dV ≈ Σ_f F_f · φ_f

    F_f = ρ·(u·n)·A (면 질량유속, 양수: owner→neighbour)

    Parameters
    ----------
    mass_flux : (n_faces,) 면 질량유속 배열
    """
    for fid, face in enumerate(mesh.faces):
        owner = face.owner
        F = mass_flux[fid]

        if face.neighbour >= 0:
            neighbour = face.neighbour
            # Upwind: max(F,0)*φ_O + min(F,0)*φ_N
            system.add_diagonal(owner, max(F, 0.0))
            system.add_off_diagonal(owner, neighbour, min(F, 0.0))
            # 이웃셀: -F (반대 방향)
            system.add_diagonal(neighbour, max(-F, 0.0))
            system.add_off_diagonal(neighbour, owner, min(-F, 0.0))
        else:
            # 경계면
            if F >= 0:
                system.add_diagonal(owner, F)
            # F < 0: 경계값이 유입 → 우변에서 처리


def temporal_operator(mesh: FVMesh, rho: float, dt: float, phi_old: np.ndarray,
                      system: FVMSystem):
    """
    시간항 이산화 (Backward Euler):
    (ρ·V/Δt)·φ = (ρ·V/Δt)·φ° + ...

    Parameters
    ----------
    rho : 밀도
    dt : 시간 간격
    phi_old : 이전 시간 스텝의 필드 값
    """
    for ci in range(mesh.n_cells):
        vol = mesh.cells[ci].volume
        coeff = rho * vol / dt
        system.add_diagonal(ci, coeff)
        system.add_source(ci, coeff * phi_old[ci])


def source_term(mesh: FVMesh, source_values: np.ndarray, system: FVMSystem):
    """
    소스항: S_P·V_P를 우변에 추가.

    Parameters
    ----------
    source_values : (n_cells,) 체적당 소스항 값
    """
    for ci in range(mesh.n_cells):
        vol = mesh.cells[ci].volume
        system.add_source(ci, source_values[ci] * vol)


def linearized_source(mesh: FVMesh, Sp: np.ndarray, Su: np.ndarray,
                      system: FVMSystem):
    """
    선형화된 소스항: S = Su + Sp·φ

    Sp < 0이어야 안정적 (음의 기울기).
    대각에 -Sp·V, 우변에 Su·V 추가.
    """
    for ci in range(mesh.n_cells):
        vol = mesh.cells[ci].volume
        system.add_diagonal(ci, -Sp[ci] * vol)
        system.add_source(ci, Su[ci] * vol)


def apply_boundary_conditions(mesh: FVMesh, phi: ScalarField, gamma: ScalarField,
                              mass_flux: np.ndarray, system: FVMSystem,
                              bc_types: dict):
    """
    경계조건 적용.

    bc_types: {patch_name: {'type': 'dirichlet'|'neumann'|'zero_gradient', ...}}
    - dirichlet: 고정값 (phi.boundary_values에서 읽음)
    - neumann: 고정 기울기 (grad 값 지정)
    - zero_gradient: ∂φ/∂n = 0
    """
    for bname, fids in mesh.boundary_patches.items():
        bc = bc_types.get(bname, {'type': 'zero_gradient'})
        bc_type = bc['type']

        for local_idx, fid in enumerate(fids):
            face = mesh.faces[fid]
            owner = face.owner
            xO = mesh.cells[owner].center
            d_Pf = np.linalg.norm(face.center - xO)

            if d_Pf < 1e-30:
                continue

            if bc_type == 'dirichlet':
                phi_b = phi.boundary_values[bname][local_idx]
                gamma_f = gamma.values[owner]
                coeff = gamma_f * face.area / d_Pf
                system.add_diagonal(owner, coeff)
                system.add_source(owner, coeff * phi_b)

                # 대류 경계 (유입)
                F = mass_flux[fid]
                if F < 0:
                    system.add_source(owner, -F * phi_b)

            elif bc_type == 'neumann':
                grad_val = bc.get('value', 0.0)
                flux = grad_val * face.area
                system.add_source(owner, flux)

            elif bc_type == 'zero_gradient':
                # 확산: 기여 없음 (0-gradient)
                # 대류: 셀 값 사용
                F = mass_flux[fid]
                if F < 0:
                    # 유입인데 zero_gradient: 셀 값 사용
                    system.add_diagonal(owner, -F)


def under_relax(system: FVMSystem, phi: ScalarField, alpha: float):
    """
    하부완화 (Under-relaxation).

    a_P/α · φ = Σ(a_N·φ_N) + b + (1-α)/α · a_P · φ_old
    """
    if alpha >= 1.0:
        return

    for ci in range(system.n):
        ap = system.diag[ci]
        system.add_diagonal(ci, ap * (1.0 - alpha) / alpha)
        system.add_source(ci, ap * (1.0 - alpha) / alpha * phi.values[ci])

File: cpp/src/fvm_operators.cpp
#include "twofluid/fvm_operators.hpp"
#include <algorithm>
#include <cmath>
#ifdef _OPENMP
#include <omp.h>
#endif

namespace twofluid {

// ---------------------------------------------------------------------------
// FVMSystem
// ---------------------------------------------------------------------------

FVMSystem::FVMSystem(int n) : n(n) {
    rhs = Eigen::VectorXd::Zero(n);
    diag = Eigen::VectorXd::Zero(n);
}

void FVMSystem::add_diagonal(int i, double val) {
    rows.push_back(i);
    cols.push_back(i);
    vals.push_back(val);
    diag(i) += val;
}

void FVMSystem::add_off_diagonal(int i, int j, double val) {
    rows.push_back(i);
    cols.push_back(j);
    vals.push_back(val);
}

void FVMSystem::add_source(int i, double val) {
    rhs(i) += val;
}

Eigen::SparseMatrix<double> FVMSystem::to_sparse() const {
    std::vector<Eigen::Triplet<double>> triplets;
    triplets.reserve(rows.size());
    for (size_t k = 0; k < rows.size(); ++k) {
        triplets.emplace_back(rows[k], cols[k], vals[k]);
    }
    Eigen::SparseMatrix<double> mat(n, n);
    mat.setFromTriplets(triplets.begin(), triplets.end());
    return mat;
}

void FVMSystem::reset() {
    rows.clear();
    cols.clear();
    vals.clear();
    rhs.setZero();
    diag.setZero();
}

// ---------------------------------------------------------------------------
// Operators
// ---------------------------------------------------------------------------

void diffusion_operator(const FVMesh& mesh, const ScalarField& gamma,
                        FVMSystem& system) {
    for (int fid = 0; fid < mesh.n_faces; ++fid) {
        const Face& face = mesh.faces[fid];
        int owner = face.owner;

        if (face.neighbour >= 0) {
            int neighbour = face.neighbour;
            Eigen::Vector3d xO = mesh.cells[owner].center;
            Eigen::Vector3d xN = mesh.cells[neighbour].center;
            double d_PN = (xN - xO).norm();
            if (d_PN < 1e-30) continue;

            double gamma_O = gamma.values(owner);
            double gamma_N = gamma.values(neighbour);
            double gamma_f = 0.0;
            if (gamma_O + gamma_N > 1e-30) {
                gamma_f = 2.0 * gamma_O * gamma_N / (gamma_O + gamma_N);
            }

            double coeff = gamma_f * face.area / d_PN;
            system.add_diagonal(owner, coeff);
            system.add_diagonal(neighbour, coeff);
            system.add_off_diagonal(owner, neighbour, -coeff);
            system.add_off_diagonal(neighbour, owner, -coeff);
        }
        // Boundary faces handled in apply_boundary_conditions
    }
}

void convection_operator_upwind(const FVMesh& mesh,
                                const Eigen::VectorXd& mass_flux,
                                FVMSystem& system) {
    for (int fid = 0; fid < mesh.n_faces; ++fid) {
        const Face& face = mesh.faces[fid];
        int owner = face.owner;
        double F = mass_flux(fid);

        if (face.neighbour >= 0) {
            int neighbour = face.neighbour;
            system.add_diagonal(owner, std::max(F, 0.0));
            system.add_off_diagonal(owner, neighbour, std::min(F, 0.0));
            system.add_diagonal(neighbour, std::max(-F, 0.0));
            system.add_off_diagonal(neighbour, owner, std::min(-F, 0.0));
        } else {
            if (F >= 0) {
                system.add_diagonal(owner, F);
            }
        }
    }
}

void temporal_operator(const FVMesh& mesh, double rho, double dt,
                       const Eigen::VectorXd& phi_old, FVMSystem& system) {
    int n = mesh.n_cells;
    // Pre-compute coefficients in parallel, then add serially (add_diagonal/add_source are not thread-safe)
    Eigen::VectorXd coeff_vec(n);
#pragma omp parallel for schedule(static)
    for (int ci = 0; ci < n; ++ci) {
        coeff_vec(ci) = rho * mesh.cells[ci].volume / dt;
    }
    for (int ci = 0; ci < n; ++ci) {
        system.add_diagonal(ci, coeff_vec(ci));
        system.add_source(ci, coeff_vec(ci) * phi_old(ci));
    }
}

void source_term(const FVMesh& mesh, const Eigen::VectorXd& source_values,
                 FVMSystem& system) {
    int n = mesh.n_cells;
    Eigen::VectorXd contrib(n);
#pragma omp parallel for schedule(static)
    for (int ci = 0; ci < n; ++ci) {
        contrib(ci) = source_values(ci) * mesh.cells[ci].volume;
    }
    for (int ci = 0; ci < n; ++ci) {
        system.add_source(ci, contrib(ci));
    }
}

void linearized_source(const FVMesh& mesh, const Eigen::VectorXd& Sp,
                       const Eigen::VectorXd& Su, FVMSystem& system) {
    int n = mesh.n_cells;
    Eigen::VectorXd diag_contrib(n), src_contrib(n);
#pragma omp parallel for schedule(static)
    for (int ci = 0; ci < n; ++ci) {
        double vol = mesh.cells[ci].volume;
        diag_contrib(ci) = -Sp(ci) * vol;
        src_contrib(ci)  =  Su(ci) * vol;
    }
    for (int ci = 0; ci < n; ++ci) {
        system.add_diagonal(ci, diag_contrib(ci));
        system.add_source(ci, src_contrib(ci));
    }
}

void apply_boundary_conditions(
    const FVMesh& mesh, const ScalarField& phi, const ScalarField& gamma,
    const Eigen::VectorXd& mass_flux, FVMSystem& system,
    const std::unordered_map<std::string, BoundaryCondition>& bc_types) {

    for (const auto& [bname, fids] : mesh.boundary_patches) {
        BoundaryCondition bc;
        bc.type = "zero_gradient";
        auto bc_it = bc_types.find(bname);
        if (bc_it != bc_types.end()) {
            bc = bc_it->second;
        }

        for (size_t local_idx = 0; local_idx < fids.size(); ++local_idx) {
            int fid = fids[local_idx];
            const Face& face = mesh.faces[fid];
            int owner = face.owner;
            Eigen::Vector3d xO = mesh.cells[owner].center;
            double d_Pf = (face.center - xO).norm();
            if (d_Pf < 1e-30) continue;

            if (bc.type == "dirichlet") {
                auto bv_it = phi.boundary_values.find(bname);
                double phi_b = (bv_it != phi.boundary_values.end())
                    ? bv_it->second(static_cast<int>(local_idx))
                    : 0.0;
                double gamma_f = gamma.values(owner);
                double coeff = gamma_f * face.area / d_Pf;
                system.add_diagonal(owner, coeff);
                system.add_source(owner, coeff * phi_b);

                double F = mass_flux(fid);
                if (F < 0) {
                    // Standard inflow (F < 0 means flow enters domain through this face).
                    // convection_operator_upwind added nothing to diagonal for this face,
                    // so we add the inflow convection contribution here.
                    system.add_source(owner, -F * phi_b);
                } else {
                    // F >= 0: convection_operator_upwind already added F to diagonal
                    // treating this as outflow. For a Dirichlet BC we must override this
                    // and enforce phi_b regardless. Cancel the diagonal outflow term and
                    // add the Dirichlet-enforced convective flux to the source.
                    system.add_diagonal(owner, -F);
                    system.add_source(owner, F * phi_b);
                }
            } else if (bc.type == "neumann") {
                double flux = bc.value * face.area;
                system.add_source(owner, flux);
            } else {  // zero_gradient
                double F = mass_flux(fid);
                if (F < 0) {
                    system.add_diagonal(owner, -F);
                }
            }
        }
    }
}

void temporal_operator_bdf2(const FVMesh& mesh, double rho, double dt,
                             const Eigen::VectorXd& phi_old,
                             const Eigen::VectorXd& phi_old_old,
                             FVMSystem& system) {
    int n = mesh.n_cells;
    Eigen::VectorXd diag_contrib(n), src_contrib(n);
#pragma omp parallel for schedule(static)
    for (int ci = 0; ci < n; ++ci) {
        double coeff = rho * mesh.cells[ci].volume / (2.0 * dt);
        diag_contrib(ci) = 3.0 * coeff;
        src_contrib(ci)  = 4.0 * coeff * phi_old(ci) - coeff * phi_old_old(ci);
    }
    for (int ci = 0; ci < n; ++ci) {
        system.add_diagonal(ci, diag_contrib(ci));
        system.add_source(ci, src_contrib(ci));
    }
}

void diffusion_operator_corrected(const FVMesh& mesh, const ScalarField& gamma,
                                   const ScalarField& phi,
                                   FVMSystem& system,
                                   int n_corrections) {
    for (int fid = 0; fid < mesh.n_faces; ++fid) {
        const Face& face = mesh.faces[fid];
        int owner = face.owner;

        if (face.neighbour >= 0) {
            int neighbour = face.neighbour;
            Eigen::Vector3d xO = mesh.cells[owner].center;
            Eigen::Vector3d xN = mesh.cells[neighbour].center;
            Eigen::Vector3d delta = xN - xO;
            double d_PN = delta.norm();
            if (d_PN < 1e-30) continue;

            double gamma_O = gamma.values(owner);
            double gamma_N = gamma.values(neighbour);
            double gamma_f = 0.0;
            if (gamma_O + gamma_N > 1e-30) {
                gamma_f = 2.0 * gamma_O * gamma_N / (gamma_O + gamma_N);
            }

            // Orthogonal distance: projection of delta onto face normal
            double d_orth = std::abs(delta.dot(face.normal));
            if (d_orth < 1e-30) d_orth = d_PN;

            // Implicit orthogonal part
            double coeff = gamma_f * face.area / d_orth;
            system.add_diagonal(owner, coeff);
            system.add_diagonal(neighbour, coeff);
            system.add_off_diagonal(owner, neighbour, -coeff);
            system.add_off_diagonal(neighbour, owner, -coeff);

            // Explicit non-orthogonal correction on RHS
            if (n_corrections > 0) {
                double phi_diff = phi.values(neighbour) - phi.values(owner);
                // cross-diffusion term: gamma_f * A_f * (phi_N - phi_O) * (1/d_orth - 1/d_PN)
                double correction = gamma_f * face.area * phi_diff
                                    * (1.0 / d_orth - 1.0 / d_PN);
                system.add_source(owner, correction);
                system.add_source(neighbour, -correction);
            }
        }
        // Boundary faces handled in apply_boundary_conditions
    }
}

void under_relax(FVMSystem& system, const ScalarField& phi, double alpha) {
    if (alpha >= 1.0) return;
    int n = system.n;
    Eigen::VectorXd factors(n);
#pragma omp parallel for schedule(static)
    for (int ci = 0; ci < n; ++ci) {
        double ap = system.diag(ci);
        factors(ci) = ap * (1.0 - alpha) / alpha;
    }
    for (int ci = 0; ci < n; ++ci) {
        system.add_diagonal(ci, factors(ci));
        system.add_source(ci, factors(ci) * phi.values(ci));
    }
}

} // namespace twofluid
